File: wear-os/app/src/main/java/com/grantlittman/wearapp/timer/TimerEngine.kt
package com.grantlittman.wearapp.timer

import com.grantlittman.wearapp.data.model.HapticType
import com.grantlittman.wearapp.data.model.Milestone
import com.grantlittman.wearapp.data.model.Pattern
import com.grantlittman.wearapp.data.model.Signal
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * The core timer logic. Tracks elapsed time and decides when to fire
 * repeating intervals, milestones, and countdown warnings.
 *
 * This class is lifecycle-agnostic — it's driven by a CoroutineScope
 * provided by the foreground service.
 */
class TimerEngine(
    private val signalExecutor: SignalExecutor,
    private val scope: CoroutineScope
) {

    companion object {
        /** Fast tick for interactive UI — smooth progress display. */
        private const val TICK_FAST_MS = 100L
        /** Slow tick for ambient/background — saves battery while still firing signals on time. */
        private const val TICK_SLOW_MS = 500L
    }

    /** Set by the activity to slow ticks when the screen is in ambient mode. */
    var isAmbient: Boolean = false

    private val _state = MutableStateFlow(TimerState())
    val state: StateFlow<TimerState> = _state.asStateFlow()

    private var tickJob: Job? = null
    private var pattern: Pattern? = null

    // Tracking what has already fired to avoid duplicates
    private var lastIntervalFiredAt: Long = 0L
    private val firedMilestoneIds = mutableSetOf<Long>() // triggerAtMillis as key
    private var lastCountdownFiredAt: Long = 0L

    /**
     * Start the timer with a given pattern.
     */
    fun start(pattern: Pattern) {
        this.pattern = pattern
        lastIntervalFiredAt = 0L
        firedMilestoneIds.clear()
        lastCountdownFiredAt = 0L

        _state.value = TimerState(
            isRunning = true,
            isPaused = false,
            elapsedMillis = 0L,
            totalDurationMillis = pattern.totalDurationMillis,
            patternName = pattern.name,
            nextMilestone = findNextMilestone(0L, pattern.milestones)
        )

        startTicking()
    }

    /**
     * Pause the timer — signals stop, elapsed time freezes.
     */
    fun pause() {
        tickJob?.cancel()
        tickJob = null
        _state.value = _state.value.copy(isRunning = false, isPaused = true)
    }

    /**
     * Resume from a paused state.
     */
    fun resume() {
        _state.value = _state.value.copy(isRunning = true, isPaused = false)
        startTicking()
    }

    /**
     * Stop the timer completely. State is preserved (not reset) so the UI
     * can still display the final values during the navigation transition.
     * The next call to start() will initialize fresh state.
     */
    fun stop() {
        tickJob?.cancel()
        tickJob = null
        signalExecutor.cancel()
        pattern = null
        _state.value = _state.value.copy(isRunning = false, isPaused = false)
    }

    private fun startTicking() {
        tickJob?.cancel()
        tickJob = scope.launch {
            var lastTickTime = System.currentTimeMillis()

            while (isActive) {
                delay(if (isAmbient) TICK_SLOW_MS else TICK_FAST_MS)

                val now = System.currentTimeMillis()
                val delta = now - lastTickTime
                lastTickTime = now

                val currentState = _state.value
                val newElapsed = currentState.elapsedMillis + delta
                val p = pattern ?: return@launch

                // Check if we've finished
                if (p.totalDurationMillis != null && newElapsed >= p.totalDurationMillis) {
                    // Fire any final milestones before stopping
                    checkMilestones(newElapsed, p)

                    // Fire a strong completion signal so the user knows time is up
                    _state.value = currentState.copy(
                        elapsedMillis = p.totalDurationMillis,
                        isRunning = false,
                        nextMilestone = null,
                        lastSignalFiredAt = System.currentTimeMillis(),
                        lastSignalWasMilestone = true
                    )
                    scope.launch {
                        signalExecutor.execute(
                            signal = Signal(
                                hapticType = HapticType.PULSE,
                                intensity = 1.0f
                            ),
                            audioEnabled = p.audioEnabled,
                            repeatCount = 3
                        )
                    }
                    return@launch
                }

                // Update elapsed time
                _state.value = currentState.copy(
                    elapsedMillis = newElapsed,
                    nextMilestone = findNextMilestone(newElapsed, p.milestones),
                    inCountdown = isInCountdown(newElapsed, p)
                )

                // Check and fire signals — milestones take priority over intervals
                val milestoneFired = checkMilestones(newElapsed, p)
                if (!milestoneFired) {
                    checkRepeatingInterval(newElapsed, p)
                }
                checkCountdownWarning(newElapsed, p)
            }
        }
    }

    /**
     * Fire the repeating interval signal if enough time has passed.
     */
    private fun checkRepeatingInterval(elapsed: Long, pattern: Pattern) {
        val interval = pattern.repeatingInterval.intervalMillis
        if (interval <= 0) return

        // How many intervals should have fired by now?
        val expectedCount = elapsed / interval
        val firedCount = lastIntervalFiredAt / interval

        if (expectedCount > firedCount) {
            lastIntervalFiredAt = expectedCount * interval
            _state.value = _state.value.copy(
                lastSignalFiredAt = System.currentTimeMillis(),
                lastSignalWasMilestone = false
            )
            scope.launch {
                signalExecutor.execute(
                    signal = pattern.repeatingInterval.signal,
                    audioEnabled = pattern.audioEnabled
                )
            }
        }
    }

    /**
     * Fire any milestones whose trigger time has been reached.
     * Returns true if at least one milestone fired (so the interval can be suppressed).
     */
    private fun checkMilestones(elapsed: Long, pattern: Pattern): Boolean {
        var fired = false
        for (milestone in pattern.milestones) {
            if (milestone.triggerAtMillis in 1..elapsed &&
                milestone.triggerAtMillis !in firedMilestoneIds
            ) {
                firedMilestoneIds.add(milestone.triggerAtMillis)
                fired = true
                _state.value = _state.value.copy(
                    lastSignalFiredAt = System.currentTimeMillis(),
                    lastSignalWasMilestone = true
                )
                scope.launch {
                    signalExecutor.execute(
                        signal = milestone.signal,
                        audioEnabled = pattern.audioEnabled,
                        repeatCount = milestone.repeatCount
                    )
                }
            }
        }
        return fired
    }

    /**
     * Fire countdown warning signals at an increased frequency during the final stretch.
     */
    private fun checkCountdownWarning(elapsed: Long, pattern: Pattern) {
        val warning = pattern.countdownWarning ?: return
        val totalDuration = pattern.totalDurationMillis ?: return

        val countdownStart = (totalDuration - warning.activateAtMillis).coerceAtLeast(0L)
        if (elapsed < countdownStart) return

        val elapsedInCountdown = elapsed - countdownStart
        val interval = warning.intervalMillis
        if (interval <= 0) return

        val lastElapsedInCountdown = (lastCountdownFiredAt - countdownStart).coerceAtLeast(0L)
        val expectedCount = elapsedInCountdown / interval
        val firedCount = lastElapsedInCountdown / interval

        if (expectedCount > firedCount) {
            lastCountdownFiredAt = countdownStart + (expectedCount * interval)
            _state.value = _state.value.copy(
                lastSignalFiredAt = System.currentTimeMillis(),
                lastSignalWasMilestone = false
            )
            scope.launch {
                signalExecutor.execute(
                    signal = warning.signal,
                    audioEnabled = pattern.audioEnabled
                )
            }
        }
    }

    private fun isInCountdown(elapsed: Long, pattern: Pattern): Boolean {
        val warning = pattern.countdownWarning ?: return false
        val totalDuration = pattern.totalDurationMillis ?: return false
        return elapsed >= (totalDuration - warning.activateAtMillis)
    }

    private fun findNextMilestone(elapsed: Long, milestones: List<Milestone>): Milestone? {
        return milestones
            .filter { it.triggerAtMillis > elapsed }
            .minByOrNull { it.triggerAtMillis }
    }
}

File: watchos/PulseTimer/PulseTimer Watch App/Timer/TimerEngine.swift
import Foundation
import Observation

/// The core timer logic. Tracks elapsed time and decides when to fire
/// repeating intervals, milestones, and countdown warnings.
///
/// UI-agnostic: drives state changes, leaves UI rendering to observers.
/// This is a direct port of Wear OS TimerEngine.kt; behavior should match exactly.
@Observable
@MainActor
final class TimerEngine {
    private static let tickFastMs: Int = 100
    private static let tickSlowMs: Int = 500

    private(set) var state: TimerState = TimerState()

    /// Set by the host (e.g., extended runtime session) to slow the tick rate when
    /// the watch is in always-on/ambient mode.
    var isAmbient: Bool = false

    private let signalExecutor: SignalExecutor
    private let runtimeController = RuntimeSessionController()
    private var tickTask: Task<Void, Never>?
    private var pattern: Pattern?

    private var lastIntervalFiredAt: Int64 = 0
    private var firedMilestoneTriggers: Set<Int64> = []
    private var lastCountdownFiredAt: Int64 = 0

    init(signalExecutor: SignalExecutor) {
        self.signalExecutor = signalExecutor
        // When the system is about to end the extended runtime session, fire a strong
        // warning haptic so the user knows the timer is about to stop tracking.
        self.runtimeController.willExpire = { [weak self] in
            guard let self else { return }
            self.signalExecutor.execute(
                signal: Signal(hapticType: .pulse, intensity: 1.0),
                audioEnabled: false,
                repeatCount: 2
            )
        }
    }

    func start(pattern: Pattern) {
        self.pattern = pattern
        lastIntervalFiredAt = 0
        firedMilestoneTriggers.removeAll()
        lastCountdownFiredAt = 0

        var newState = TimerState()
        newState.isRunning = true
        newState.elapsedMillis = 0
        newState.totalDurationMillis = pattern.totalDurationMillis
        newState.patternName = pattern.name
        newState.nextMilestone = Self.findNextMilestone(elapsed: 0, in: pattern.milestones)
        state = newState

        runtimeController.start()
        startTicking()
    }

    func pause() {
        tickTask?.cancel()
        tickTask = nil
        state.isRunning = false
        state.isPaused = true
    }

    func resume() {
        state.isRunning = true
        state.isPaused = false
        startTicking()
    }

    /// Stop completely. State is preserved (not reset) so the UI can keep displaying
    /// final values during the navigation transition. The next `start()` reinitializes.
    func stop() {
        tickTask?.cancel()
        tickTask = nil
        signalExecutor.cancel()
        runtimeController.stop()
        pattern = nil
        state.isRunning = false
        state.isPaused = false
    }

    private func startTicking() {
        tickTask?.cancel()
        tickTask = Task { @MainActor [weak self] in
            var lastTickTime = Self.nowMillis()
            while !Task.isCancelled {
                guard let self else { return }
                let intervalMs = self.isAmbient ? Self.tickSlowMs : Self.tickFastMs
                do {
                    try await Task.sleep(for: .milliseconds(intervalMs))
                } catch {
                    return
                }
                let now = Self.nowMillis()
                let delta = now - lastTickTime
                lastTickTime = now
                self.tick(delta: delta)
            }
        }
    }

    private func tick(delta: Int64) {
        guard let pattern = self.pattern else { return }
        let newElapsed = state.elapsedMillis + delta

        // Reaching total duration ends the session with a final completion signal.
        if let total = pattern.totalDurationMillis, newElapsed >= total {
            _ = checkMilestones(elapsed: newElapsed, pattern: pattern)
            state.elapsedMillis = total
            state.isRunning = false
            state.nextMilestone = nil
            state.lastSignalFiredAt = Self.nowMillis()
            state.lastSignalWasMilestone = true
            // Single PULSE for completion. Wear OS fires 3, but on Apple Watch the
            // haptic motor sits under the display and back-to-back strong pulses
            // make the Done button briefly unresponsive to taps. One pulse plus the
            // visible "Done!" UI is more than enough as a completion cue.
            signalExecutor.execute(
                signal: Signal(hapticType: .pulse, intensity: 1.0),
                audioEnabled: pattern.audioEnabled,
                repeatCount: 1
            )
            tickTask?.cancel()
            tickTask = nil
            // Intentionally NOT ending the runtime session here. We need it to stay
            // alive through the "Done!" screen so the app remains foreground-active
            // and the Done button responds to the first tap. If we end the session
            // here, the screen sleeps after a few idle seconds and the next tap
            // wakes the app instead of registering the dismissal — making the Done
            // button feel stuck. Session is properly ended in `stop()` when the user
            // taps Done.
            return
        }

        state.elapsedMillis = newElapsed
        state.nextMilestone = Self.findNextMilestone(elapsed: newElapsed, in: pattern.milestones)
        state.inCountdown = Self.isInCountdown(elapsed: newElapsed, pattern: pattern)

        // Milestones take priority over the regular interval on the same tick.
        let milestoneFired = checkMilestones(elapsed: newElapsed, pattern: pattern)
        if !milestoneFired {
            checkRepeatingInterval(elapsed: newElapsed, pattern: pattern)
        }
        checkCountdownWarning(elapsed: newElapsed, pattern: pattern)
    }

    private func checkRepeatingInterval(elapsed: Int64, pattern: Pattern) {
        let interval = pattern.repeatingInterval.intervalMillis
        guard interval > 0 else { return }
        let expectedCount = elapsed / interval
        let firedCount = lastIntervalFiredAt / interval
        guard expectedCount > firedCount else { return }
        lastIntervalFiredAt = expectedCount * interval
        state.lastSignalFiredAt = Self.nowMillis()
        state.lastSignalWasMilestone = false
        signalExecutor.execute(
            signal: pattern.repeatingInterval.signal,
            audioEnabled: pattern.audioEnabled
        )
    }

    @discardableResult
    private func checkMilestones(elapsed: Int64, pattern: Pattern) -> Bool {
        var fired = false
        for milestone in pattern.milestones {
            let trigger = milestone.triggerAtMillis
            guard trigger >= 1, trigger <= elapsed,
                  !firedMilestoneTriggers.contains(trigger) else { continue }
            firedMilestoneTriggers.insert(trigger)
            fired = true
            state.lastSignalFiredAt = Self.nowMillis()
            state.lastSignalWasMilestone = true
            signalExecutor.execute(
                signal: milestone.signal,
                audioEnabled: pattern.audioEnabled,
                repeatCount: milestone.repeatCount
            )
        }
        return fired
    }

    private func checkCountdownWarning(elapsed: Int64, pattern: Pattern) {
        guard let warning = pattern.countdownWarning,
              let total = pattern.totalDurationMillis else { return }
        let countdownStart = max(0, total - warning.activateAtMillis)
        guard elapsed >= countdownStart else { return }

        let elapsedInCountdown = elapsed - countdownStart
        let interval = warning.intervalMillis
        guard interval > 0 else { return }

        let lastElapsedInCountdown = max(0, lastCountdownFiredAt - countdownStart)
        let expectedCount = elapsedInCountdown / interval
        let firedCount = lastElapsedInCountdown / interval
        guard expectedCount > firedCount else { return }

        lastCountdownFiredAt = countdownStart + (expectedCount * interval)
        state.lastSignalFiredAt = Self.nowMillis()
        state.lastSignalWasMilestone = false
        signalExecutor.execute(
            signal: warning.signal,
            audioEnabled: pattern.audioEnabled
        )
    }

    private static func isInCountdown(elapsed: Int64, pattern: Pattern) -> Bool {
        guard let warning = pattern.countdownWarning,
              let total = pattern.totalDurationMillis else { return false }
        return elapsed >= (total - warning.activateAtMillis)
    }

    private static func findNextMilestone(elapsed: Int64, in milestones: [Milestone]) -> Milestone? {
        milestones
            .filter { $0.triggerAtMillis > elapsed }
            .min { $0.triggerAtMillis < $1.triggerAtMillis }
    }

    private static func nowMillis() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}
